server/src/main/java/nl/inl/blacklab/server/search/ResultsCache.java

```diff
@@ -83,27 +87,62 @@ public T get() throws InterruptedException, ExecutionException {
         public T get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
             return results;
         }
+
+    }
+
+    private static final class SearchInfoWrapper {
+        private final Search<? extends SearchResult>  search;
+        private final String requestId;
+
+        public SearchInfoWrapper(Search<? extends SearchResult> search, String requestId) {
+            this.search = search;
+            this.requestId = requestId;
+        }
+
+        public Search<? extends SearchResult> getSearch() {
+            return search;
+        }
+
+        public String getRequestId() {
+            return requestId;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()){
+                return false;
+            }
+            SearchInfoWrapper that = (SearchInfoWrapper) o;
+            return search.equals(that.search);
+        }
+
+        @Override
+        public int hashCode() {
+            return search.hashCode();
+        }
     }
 
     public ResultsCache(BLSConfig config, ExecutorService threadPool, LogDatabase logDatabase)  {
         this.threadPool = threadPool;
 
-        CacheLoader<Search<? extends SearchResult>, SearchResult> cacheLoader = new CacheLoader<Search<? extends SearchResult>, SearchResult>() {
+        CacheLoader<SearchInfoWrapper, SearchResult> cacheLoader = new CacheLoader<SearchInfoWrapper, SearchResult>() {
             @Override
-            public @Nullable SearchResult load(Search<?> search) throws Exception {
-                ThreadContext.put("requestId", Integer.toHexString(search.hashCode()));
-                long start = System.currentTimeMillis();
-                Future<? extends SearchResult> job;
-                if (runningJobs.containsKey(search)) {
-                    job = runningJobs.get(search);
-                } else {
-                    job = ResultsCache.this.threadPool.submit(search::executeInternal);
-                    runningJobs.put(search, job);
-                }
-                SearchResult searchResult = job.get();
-                logger.warn("Search time is: {}", System.currentTimeMillis() - start);
-                runningJobs.remove(search);
-                return searchResult;
+            public SearchResult load(final SearchInfoWrapper searchWrapper) throws Exception {
+                final String requestId = searchWrapper.getRequestId();
+                ThreadContext.put("requestId", requestId);
+                Future<CacheEntryWithResults<? extends SearchResult>> job = runningJobs.computeIfAbsent(searchWrapper.getSearch(), (search) -> ResultsCache.this.threadPool.submit(() -> {
```

Comment:
Should we just use caffeine's async cache loading to avoid having to get this right ourselves?